Machinery to resolve write intents during the course of inconsistent rea...

spencerkimball opened the pull request:
...ds.

Previously, inconsistent reads would simply ignore write intents. This is
not correct for two reasons. The biggest is that we do range lookups
with inconsistent reads. If a dangling intent is left on a meta record,
this means that we'd never have a way to get to the up-to-date meta
record, which might contain the latest information about a split, merge
or replica change. The second, smaller reason is that keys which are
only read inconsistently (say for performance) would have to wait for
a GC event in order to be resolved, which could be a long time.

This change does two things:
- Changes MVCC to continue a scan in the face of write intent errors
  and batch up all of them into a list of intents, to be resolved in
  parallel. Inconsistent reads continue, as before, in spite of intents,
  but now return a write intent error containing all intents encountered.
- At the store level, we now formulate an internal batch command to
  do this efficiently.

There's a big caveat which is documented in the code related to infinte
loops caused by intents encountered on range meta records. The basic
gist is that the intent is encountered on range lookup, which then
causes the store to push the txn, which causes the distributed sender
to lookup the txn. Since the txn record is always located on the range
which the meta record with the intent points to, this would in turn
cause another write intent error, and the cycle would repeat. The
solution to this is hairy and makes me nervous, but I couldn't find
any less obnoxious way.

TODO: a unittest which does the following:
- Split a range
- Ignore intent resolves for meta records
- Send a get to the second half of the range; this would normally
  cause the infinite attempts to resolve.

spencerkimball committed:
Machinery to resolve write intents during the course of inconsistent reads.

Previously, inconsistent reads would simply ignore write intents. This is
not correct for two reasons. The biggest is that we do range lookups
with inconsistent reads. If a dangling intent is left on a meta record,
this means that we'd never have a way to get to the up-to-date meta
record, which might contain the latest information about a split, merge
or replica change. The second, smaller reason is that keys which are
only read inconsistently (say for performance) would have to wait for
a GC event in order to be resolved, which could be a long time.

This change does two things:

- Changes MVCC to continue a scan in the face of write intent errors
  and batch up all of them into a list of intents, to be resolved in
  parallel. Inconsistent reads continue, as before, in spite of intents,
  but now return a write intent error containing all intents encountered.
- At the store level, we now formulate an internal batch command to
  do this efficiently.

There's a big caveat which is documented in the code related to infinte
loops caused by intents encountered on range meta records. The basic
gist is that the intent is encountered on range lookup, which then
causes the store to push the txn, which causes the distributed sender
to lookup the txn. Since the txn record is always located on the range
which the meta record with the intent points to, this would in turn
cause another write intent error, and the cycle would repeat. The
solution to this is hairy and makes me nervous, but I couldn't find
any less obnoxious way.

spencerkimball commented:
@bdarnell

tschottdorf commented on storage/range.go:
```diff
@@ -1252,6 +1259,8 @@ func (r *Range) InternalRangeLookup(batch engine.Engine, args *proto.InternalRan
+	} else if args.IgnoreIntents {
+		rangeCount = 1 // simplify lookup because we may have to retry to read new
 	}
 
```
would add something to the description of InternalRangeLookup that hints at the fact that there's a special consistency handling here and that ReadConsistency is ignored.

spencerkimball replied:
Done

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -496,11 +503,17 @@ func (ds *DistSender) Send(call client.Call) {
 		args.Header().Timestamp = ds.clock.Now()
 	}
 
+	// If this is an InternalPushTxn, set ignoreIntents option as necessary.
```
`+ // This prevents a potential infinite loop; see the comments in proto.InternalLookupRangeRequest.`

spencerkimball replied:
Done.

spencerkimball committed:
Rebase and added comment

tschottdorf commented on proto/errors.proto:
```diff
@@ -91,17 +91,21 @@ message TransactionStatusError {
-  optional bytes key = 1 [(gogoproto.nullable) = false, (gogoproto.customtype) = "Key"];
-  optional Transaction txn = 2 [(gogoproto.nullable) = false];
-  optional bool resolved = 3 [(gogoproto.nullable) = false];
+  message Intent {
```
does it make any difference having this nested inside of WriteIntentError other than that you'd have to refer to it as WriteIntentError.Intent if you were to use it in another message? Just curious.

spencerkimball replied:
It just keeps the namespace cleaner by not littering it with an "Intent" object. You need to refer to these as `WriteIntentError_Intent`.

tschottdorf commented on storage/engine/mvcc.go:
```diff
@@ -480,11 +480,12 @@ func mvccGetInternal(engine Engine, key proto.Key, metaKey proto.EncodedKey, tim
-		if !timestamp.Less(meta.Timestamp) {
-			timestamp = meta.Timestamp.Prev()
-		}
+	// at is a historical timestamp < the intent timestamp. However, we
```
so the caller will get a value and a WriteIntentError, and it basically needs to know that because it was reading inconsistently, the value is the legitimate result of the inconsistent read?

tschottdorf replied:
Probably a good time to explain the `consistent` param in the comment to `mvccGetInternal`.

spencerkimball replied:
Yes, I added more comments.

tschottdorf commented on storage/engine/mvcc.go:
```diff
@@ -497,7 +498,7 @@ func mvccGetInternal(engine Engine, key proto.Key, metaKey proto.EncodedKey, tim
 		if meta.Txn != nil && (txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID)) {
```
this part could be untangled a little bit now that we're also possibly returning the same error in other cases. But let's leave it for this PR.

bdarnell commented on kv/dist_sender_test.go:
```diff
@@ -305,6 +305,36 @@ func TestRetryOnNotLeaderError(t *testing.T) {
+		rpcSend: testFn,
+		rangeDescriptorDB: mockRangeDescriptorDB(func(_ proto.Key, opts lookupOptions) ([]proto.RangeDescriptor, error) {
+			if !opts.ignoreIntents {
+				t.Fatal("expected ignore intents to be true")
```
Should we also have a test that ignoreIntents is not true in other cases?

spencerkimball replied:
Added.

bdarnell commented on kv/dist_sender.go:
```diff
@@ -252,6 +252,11 @@ func (ds *DistSender) verifyPermissions(args proto.Request) error {
 }
 
+// lookupOptions capture additional options to pass to InternalRangeLookup.
+type lookupOptions struct {
```
Do you anticipate other options to be added to this struct in the future? Although I actually kind of like the struct even if it's a single bool; it makes the code more self-documenting than a plain true/false.

spencerkimball replied:
I don't have anything in mind, but threading a random bool through the range descriptor stuff was a heinous crime, so I opted for this.

bdarnell commented on storage/engine/mvcc.go:
```diff
@@ -1353,6 +1383,7 @@ func MVCCComputeStats(engine Engine, key, endKey proto.Key, nowNanos int64) (pro
 			}
 			ms.KeyBytes += int64(len(kv.Key))
 			ms.ValBytes += int64(len(kv.Value))
+			log.Infof("val %s: %d", kv.Key, len(kv.Value))
```
Remove these logs.

tschottdorf replied:
remove.

spencerkimball replied:
Yeah, trying to find a heisenbug. Will remove before submission.

bdarnell commented on storage/range_command.go:
```diff
@@ -328,6 +328,10 @@ func (r *Range) EndTransaction(batch engine.Engine, ms *proto.MVCCStats, args *p
 // by their current workload.
+//
+// Note that read consistency specified in the args header is ignored. Reads
+// are always inconsistent, but every attempt is made to resolve encountered
```
In #852 I need a way to do a consistent InternalRangeLookup.

spencerkimball replied:
OK, made the read consistency configurable again. This means that a request to consistent read a range meta record will hit the intent, try to push the txn with RangeLookup=true, do an inconsistent range lookup to find txn, hit intent, and return random old or new value. It will then try to push the txn. If it fails, it will backoff and retry again until it succeeds.

tschottdorf commented on storage/engine/mvcc.go:
```diff
@@ -939,14 +947,23 @@ func MVCCIterate(engine Engine, key, endKey proto.Key, timestamp proto.Timestamp
 	}
 
+	// A cumulative write intent error to gather all write intents.
+	var wiErr error
```
s/error/*proto.WriteIntentError/

spencerkimball replied:
Actually this is intentional and a byproduct of `var wiErr *proto.WriteIntentError`, when returned as an `error` not being `nil` even if it is nil.

tschottdorf replied:
What do you mean? http://play.golang.org/p/hDFdEcwlKZ

bdarnell replied:
A nil interface value is different from a typed nil pointer: http://golang.org/doc/faq#nil_error

http://play.golang.org/p/UVlgMpAxHL

tschottdorf replied:
oh, I see.

tschottdorf commented on storage/engine/mvcc.go:
```diff
@@ -957,12 +974,25 @@ func MVCCIterate(engine Engine, key, endKey proto.Key, timestamp proto.Timestamp
+				if wiErr == nil {
+					wiErr = t
+				} else {
+					wiErr.(*proto.WriteIntentError).Intents = append(wiErr.(*proto.WriteIntentError).Intents, t.Intents...)
```
no interface conversion needed if you just start out with `wiErr` as suggested above.

spencerkimball replied:
See above note.

bdarnell commented on storage/range_command.go:
```diff
@@ -339,6 +343,8 @@ func (r *Range) InternalRangeLookup(batch engine.Engine, args *proto.InternalRan
 		reply.SetGoError(util.Errorf(
 			"Range lookup specified invalid maximum range count %d: must be > 0", rangeCount))
 		return
+	} else if args.IgnoreIntents {
```
Remove the `else`. It's unnecessary after the return in the `if`, and it's harder to follow when the structure implies that `args.IgnoreIntents` and `rangeCount < 1` are somehow related.

spencerkimball replied:
OK.

tschottdorf commented on storage/engine/mvcc.go:
```diff
@@ -1384,6 +1415,7 @@ func MVCCComputeStats(engine Engine, key, endKey proto.Key, nowNanos int64) (pro
 			}
 			ms.KeyBytes += mvccVersionTimestampSize
 			ms.ValBytes += int64(len(kv.Value))
+			log.Infof("val %s @%s: %d", kv.Key, ts, len(kv.Value))
```
ditto

tschottdorf commented on storage/range.go:
```diff
@@ -515,13 +515,20 @@ func (r *Range) SetLastVerificationTimestamp(timestamp proto.Timestamp) error {
 // Raft without waiting for their completion.
 func (r *Range) AddCmd(args proto.Request, reply proto.Response, wait bool) error {
+	header := args.Header()
+	if !r.ContainsKeyRange(header.Key, header.EndKey) {
```
where's the usual check for this?

spencerkimball replied:
Post raft.

tschottdorf commented on storage/range_command.go:
```diff
@@ -346,16 +352,39 @@ func (r *Range) InternalRangeLookup(batch engine.Engine, args *proto.InternalRan
 
 	// The initial key must have the same metadata level prefix as we queried.
-	if len(kvs) == 0 {
+	if err == nil && len(kvs) == 0 {
```
this might be a rare case where an if/else with the previous block is the better option.

spencerkimball replied:
I folded this into the if / else statement above.

bdarnell commented on storage/range_command.go:
```diff
@@ -346,16 +352,39 @@ func (r *Range) InternalRangeLookup(batch engine.Engine, args *proto.InternalRan
+			// the range may or may not have changed depending on the success of
+			// the split/merge/update, we return the range info without intent
+			// errors. However, because we don't know whether the range update
+			// succeeded or failed, we are forced to return old or new randomly.
```
I find this comment unclear; how about:

On the range-lookup from a push transaction, we therefore want to suppress WriteIntentErrors and return a value anyway. But which value? We don't know whether the range update succeeded or failed, but if we don't return the correct range descriptor we may not be able to find the transaction to push. Since we cannot know the correct answer, we choose randomly between the pre- and post- transaction values. If we guess wrong, the client will try again and get the other value (within a few tries).

spencerkimball replied:
OK.

tschottdorf commented on storage/range_test.go:
```diff
@@ -2194,6 +2194,82 @@ func TestChangeReplicasDuplicateError(t *testing.T) {
+			t.Errorf("expected orig/new descriptor %s/%s; got %s", origDesc, newDesc, rlReply.Ranges[0])
+		}
+	}
+	if origCount > newCount+20 || origCount < newCount-20 {
```
Kind of awkward to test something random like this, but I suppose it'll never fail this test while we look at it, so it's ok. I think I'd still prefer a parity instead of the rand in the range lookup (you could just have a package-local count and `(c+1)%2` every time you would normally roll the dice.

spencerkimball replied:
I think random is a stronger guarantee. You could imagine dueling requests which continuously arrive, one after the other both incrementing the count, neither ever succeeding. It's ridiculously unlikely, but then again, possible. And random doesn't suffer from this failure scenario.

tschottdorf replied:
I mostly find the test slightly awkward. But I can live with that.

tschottdorf commented on storage/store.go:
```diff
@@ -1005,12 +1007,25 @@ func (s *Store) ExecuteCmd(args proto.Request, reply proto.Response) error {
+					header.ReadConsistency = proto.CONSISTENT
+				}
 				return util.RetryReset, nil
+			} else if header.ReadConsistency == proto.INCONSISTENT {
```
remove the `else` since you returned from the previous branch.

spencerkimball replied:
Good point. Done.

tschottdorf commented on storage/store.go:
```diff
@@ -1005,12 +1007,25 @@ func (s *Store) ExecuteCmd(args proto.Request, reply proto.Response) error {
+			for _, intent := range t.Intents {
+				if proto.IsWrite(args) && header.Timestamp.Less(intent.Txn.Timestamp) {
+					header.Timestamp = intent.Txn.Timestamp
+					header.Timestamp.Logical++
```
use `Next()`.

tschottdorf replied:
I mean `header.Timestamp.Forward(intent.Txn.Timestamp.Next())`.

spencerkimball replied:
Nice. Done.

bdarnell commented on storage/range_command.go:
```diff
@@ -346,16 +352,39 @@ func (r *Range) InternalRangeLookup(batch engine.Engine, args *proto.InternalRan
+			// the split/merge/update, we return the range info without intent
+			// errors. However, because we don't know whether the range update
+			// succeeded or failed, we are forced to return old or new randomly.
+			if rand.Intn(2) == 0 {
```
In the event of a retry, will the same timestamp be used or will it be changed? If retries get a new timestamp then we could use `hash(timestamp)%2` instead of a random number to avoid randomness in tests.

The important thing is that we return a descriptor that includes the current leader, right? (errors in the other direction are less critical, since we must be tolerant of seeing non-leaders in the descriptor). For ChangeReplicas, we could avoid randomness completely: simply return the descriptor with the longer replica list (since all changes are either the addition or removal of one node). I'm not sure if a similar logic could apply for splits and merges, but this might be a way to avoid non-determinism here.

spencerkimball replied:
Retries use the same timestamp. The non-determinism in the test is well-contained. This is not a problem which needs a solution, IMO.

tschottdorf replied:
ok. You could use the hlc clock instead, reading off the milliseconds as your rand and control that nicely in tests. But... your code, your call.

bdarnell commented on storage/store_test.go:
```diff
@@ -66,14 +66,58 @@ type testSender struct {
+
 // Send forwards the call to the single store. This is a poor man's
 // version of kv/coordinator, but it serves the purposes of supporting
 // tests in this package. Batches and transactions are not
```
s/Batches and t/T/

spencerkimball replied:
Done.

bdarnell commented on storage/store_test.go:
```diff
@@ -66,14 +66,58 @@ type testSender struct {
+}
+
 // Send forwards the call to the single store. This is a poor man's
 // version of kv/coordinator, but it serves the purposes of supporting
```
s/coordinator/txn_coord_sender/

spencerkimball replied:
Done.

tschottdorf commented on storage/store_test.go:
```diff
@@ -1003,62 +1047,181 @@ func TestStoreResolveWriteIntentNoTxn(t *testing.T) {
+		if err := store.ExecuteCmd(gArgs, gReply); err != nil {
+			t.Errorf("expected read to succeed: %s", err)
+		}
+		// The new value of B will be read as we can push txn B.
```
this should also resolve the intent on B (though asynchronously), right? Are you checking that?

tschottdorf replied:
oh, I think the next test is what I was looking for.

tschottdorf commented on storage/store_test.go:
```diff
@@ -1003,62 +1047,181 @@ func TestStoreResolveWriteIntentNoTxn(t *testing.T) {
+
+		wait := 1 * time.Second
+		if !test.expFinish {
+			wait = 10 * time.Millisecond
```
can you not control the retryopts and get the info that way?

bdarnell commented:
LGTM, although as we just discussed I think we need to make the call to resolveWriteIntent asynchronous when it was triggered from an inconsistent read.

tschottdorf commented:
LGTM

spencerkimball committed:
Address PR feedback.

- Handle resolves on inconsistent reads asynchronously.
- Added new push transaction action: CONFIRM_NOT_PENDING, because we
  don't want inconsistent reads to be pushing other txns unless they're
  missing a heartbeat, not pending anymore, or of a different epoch.
- Allow consistent range lookup.

spencerkimball commented:
Had to make some major changes. See newest commit. PTAL

spencerkimball committed:
Incorporate feedback.

Also removed transaction push check for an earlier txn epoch. Just because
an intent from an earlier epoch is encountered doesn't mean the txn can be
pushed. This was just incorrect. Removed associated unittest.

spencerkimball committed:
Additional feedback.
